rust-src/wallet_library/src/statement.rs

```diff
@@ -0,0 +1,375 @@
+};
+use std::{
+    collections::{BTreeMap, HashSet},
+    fmt::Display,
+    hash::Hash,
+    marker::PhantomData,
+};
+
+pub trait AcceptableRequest<C: Curve, AttributeType: Attribute<C::Scalar>> {
+    /// Method to check whether the request is acceptable given the provided
+    /// config. i.e. that it satifies the rules in the config, and some
+    /// general rules.
+    fn acceptable_request(
+        &self,
+        config: &WalletConfig<C, AttributeType>,
+    ) -> AcceptableRequestResult;
+}
+
+pub trait AcceptableAtomicStatement<
+    C: Curve,
+    TagType: AttributeTagType,
+    AttributeType: Attribute<C::Scalar>,
+> {
+    /// Method to check that the statement is acceptable given the provided
+    /// rules.
+    fn acceptable_atomic_statement(
+        &self,
+        rules: &Option<WalletConfigRules<C, TagType, AttributeType>>,
+    ) -> AcceptableRequestResult;
+}
+
+pub type AcceptableRequestResult = Result<(), RequestCheckError>;
+
+pub struct WalletConfig<'a, C: Curve, AttributeType: Attribute<C::Scalar>> {
+    /// Rules that statements on identity credentials should satisfy
+    pub identity_rules: Option<WalletConfigRules<'a, C, AttributeTag, AttributeType>>,
+    /// Rules that statements on web3Id credentials should satisfy. If a
+    /// statement uses a contract index without an entry, it should only
+    /// consider for basic, global, rules.
+    pub web3_rules:     BTreeMap<ContractAddress, WalletConfigRules<'a, C, String, AttributeType>>,
+}
+
+pub struct WalletConfigRules<
+    'a,
+    C: Curve,
+    TagType: AttributeTagType,
+    AttributeType: Attribute<C::Scalar>,
+> {
+    /// The set of tags, which are allowed to be used for range statements.
+    pub range_tags:      HashSet<TagType>,
+    /// The set of tags, which are allowed to be used for membership and
+    /// nonMembership statements.
+    pub set_tags:        HashSet<TagType>,
+    /// A function to check attributes using custom behaviour.
+    pub attribute_check: AttributeCheck<'a, TagType, AttributeType>,
+    pub _marker:         PhantomData<C>,
+}
+
+pub trait AttributeTagType: Serialize + Ord + Hash + Display {}
+impl<T: Serialize + Ord + Hash + Display> AttributeTagType for T {}
+
+type AttributeCheck<'a, TagType, AttributeType> =
+    Box<dyn Fn(&TagType, &AttributeType) -> AcceptableRequestResult + 'a>;
+
+#[derive(Debug, thiserror::Error, Clone)]
+pub enum RequestCheckError {
+    #[error("Credential statement must include atomic statements")]
+    EmptyCredentialStatement,
+    #[error("Credential statement may not reuse attribute tags across atomic statements")]
+    DuplicateTag,
+    #[error("Web3Id statement support have not been added yet")]
+    Web3IdStatementNotSupported,
+    #[error("`{0}`")]
+    InvalidValue(String),
+    #[error("Range statement min must be less than max")]
+    RangeMinMaxError,
+    #[error("The tag `{0}` is not allowed to be used for range statements")]
+    IllegalRangeTag(String),
```

Comment:
I don't quite get why this is a String since this loses information. Why not just put the tag here directly?

Reply:
The tag can be a string (for web3Id), so this would need to be generic, so I thought making this a string was preferable.